test/core/test_lwt.ml

```diff
@@ -2720,6 +2720,94 @@ let protected_tests = suite "protected" [
 ]
 let suites = suites @ [protected_tests]
 
+let cancelable_tests = suite "wrap_in_cancelable" [
+  test "fulfilled" begin fun () ->
+    let p = Lwt.wrap_in_cancelable (Lwt.return ()) in
+    Lwt.return (Lwt.state p = Lwt.Return ())
+  end;
+
+  test "rejected" begin fun () ->
+    let p = Lwt.protected (Lwt.fail Exception) in
+    Lwt.return (Lwt.state p = Lwt.Fail Exception)
+  end;
+
+  test "pending(task)" begin fun () ->
+    let p, _ = Lwt.task () in
+    let p' = Lwt.wrap_in_cancelable p in
+    Lwt.return (Lwt.state p = Lwt.Sleep && Lwt.state p' = Lwt.Sleep)
+  end;
+
+  test "pending(task), fulfilled" begin fun () ->
+    let p, r = Lwt.task () in
+    let p' = Lwt.wrap_in_cancelable p in
+    Lwt.wakeup r "foo";
+    Lwt.return (Lwt.state p = Lwt.Return "foo" && Lwt.state p' = Lwt.Return "foo")
+  end;
+
+  test "pending(task), canceled" begin fun () ->
+    let p, _ = Lwt.task () in
+    let p' = Lwt.wrap_in_cancelable p in
+    Lwt.cancel p';
+    Lwt.return (Lwt.state p = Lwt.Fail Lwt.Canceled && Lwt.state p' = Lwt.Fail Lwt.Canceled)
+  end;
+
+  test "pending(wait)" begin fun () ->
+    let p, _ = Lwt.wait () in
+    let p' = Lwt.wrap_in_cancelable p in
+    Lwt.return (Lwt.state p = Lwt.Sleep && Lwt.state p' = Lwt.Sleep)
+  end;
+
+  test "pending(wait), fulfilled" begin fun () ->
+    let p, r = Lwt.wait () in
+    let p' = Lwt.wrap_in_cancelable p in
+    Lwt.wakeup r "foo";
+    Lwt.return (Lwt.state p = Lwt.Return "foo" && Lwt.state p' = Lwt.Return "foo")
+  end;
+
+  test "pending(wait), canceled" begin fun () ->
+    let p, _ = Lwt.wait () in
+    let p' = Lwt.wrap_in_cancelable p in
+    Lwt.cancel p';
+    Lwt.return (Lwt.state p = Lwt.Sleep && Lwt.state p = Lwt.Fail Lwt.Canceled)
+  end;
+
+  test "pending(task), canceled, fulfilled" begin fun () ->
+    let p, r = Lwt.task () in
+    let p' = Lwt.protected p in
+    Lwt.cancel p';
+    Lwt.wakeup r "foo";
+    Lwt.return
+      (Lwt.state p = Lwt.Fail Lwt.Canceled && Lwt.state p' = Lwt.Fail Lwt.Canceled)
```

Comment:
What causes `p` to become canceled here? IIRC `protected` should prevent propagation of cancellation from `p'` to `p`.

Reply:
Yes, this should be `Return "foo"`. I changed it now.

Also, it doesn't actually test the added function at all. And several of those tests have the same issue. I'll fix that too.